prow/plugins/projectmanager/projectmanager.go

```diff
@@ -0,0 +1,200 @@
+	// Only respond to label add or issue/PR open events
+	if pe.Action != github.PullRequestActionOpened &&
+		pe.Action != github.PullRequestActionReopened &&
+		pe.Action != github.PullRequestActionLabeled {
+		return nil
+	}
+	// Get any ManagedProjects that match this PR
+	matchedColumnIDs, err := getMatchingColumnIDs(gc, projectManager, pe)
+	if err != nil {
+		return err
+	}
+	// For each ManagedColumn that matches this PR, add this PR to that Project Column
+	for _, matchedColumnID := range matchedColumnIDs {
+		err = addPRToColumn(gc, matchedColumnID, pe)
+		log.WithError(err).Println("Failed to add PR to project")
+	}
+	return nil
+}
+
+func getMatchingColumnIDs(gc githubClient, projectManager plugins.ProjectManager, pe github.PullRequestEvent) ([]int, error) {
+	var matchedColumnIDs []int
+	// Don't use GetIssueLabels unless it's required and keep track of whether the labels have been fetched to avoid unnecessary API usage.
+	labelsFetched := false
+	var labels []github.Label
+	for orgRepoName, managedOrgRepo := range projectManager.OrgRepos {
+		for projectName, managedProject := range managedOrgRepo.Projects {
+			for _, managedColumn := range managedProject.Columns {
+				if managedColumn.Org != "" && managedColumn.Org != pe.Repo.Owner.Login {
+					continue
+				}
+				if managedColumn.State != "" && managedColumn.State != pe.PullRequest.State {
+					continue
+				}
+				if len(managedColumn.Labels) != 0 {
+					if !labelsFetched {
+						// If labels are not yet fetched then get them as they are now required
+						// GetIssueLabels works for PRs as they are considered issues in the API
+						var err error
+						labels, err = gc.GetIssueLabels(pe.Repo.Owner.Login, pe.Repo.Name, pe.Number)
+						if err != nil {
+							return nil, err
+						}
+						labelsFetched = true
+					}
+					if !github.HasLabels(managedColumn.Labels, labels) {
+						continue
+					}
+				}
+				columnID := managedColumn.ID
+				// Currently this assumes columnID having a value if 0 means it is unset
+				// While it's highly unlikely that an actual project would have an ID of 0, given that
+				// these IDs are global across GitHub, this doesn't seem like an ideal solution.
+				if columnID == nil {
+					var err error
+					columnID, err = getColumnID(orgRepoName, projectName, managedColumn.Name, gc)
+					if err != nil {
+						return nil, err
+					}
+				}
+				matchedColumnIDs = append(matchedColumnIDs, *columnID)
+			}
+		}
+	}
+	return matchedColumnIDs, nil
+}
+
+func getColumnID(orgRepoName, projectName, columnName string, gc githubClient) (*int, error) {
+	var projects []github.Project
+	var err error
+	orgRepoParts := strings.Split(orgRepoName, "/")
+	switch len(orgRepoParts) {
+	case 2:
+		projects, err = gc.GetRepoProjects(orgRepoParts[0], orgRepoParts[1])
+	case 1:
+		projects, err = gc.GetOrgProjects(orgRepoParts[0])
+	default:
+		return nil, fmt.Errorf("could not determine org or org/repo from %s", orgRepoName)
```

Comment:
Ideally we would include this validation in the plugin config parsing logic so that this fails when config is loaded instead of when we try to use the value.

Reply:
I think the yaml cannot be valid if org/repo is not specified,  case 1 or case 2 are only possibilities. I would also like to know if there is a callback to validate the configuration within plugin.